ArrayList/ArrayList.cs

```diff
@@ -1,8 +1,580 @@
+
+                throw new IndexOutOfRangeException();
+            }
+            set
+            {
+                if (index >= 0 && index < _array.Length)
+                {
+                    _array[index] = value;
+                }
+
+                throw new IndexOutOfRangeException();
+            }
+        }
+
+        private int _length;
+        private T[] _array;
+        private int _initLength = 10;
+
+        public ArrayList()
+        {
+            Length = 0;
+            _array = new T[_initLength];
+        }
+
+        public ArrayList(T value)
+        {
+            if (value != null)
+            {
+                Length = 1;
+                _array = new T[_initLength];
+                _array[0] = value;
+            }
+            else
+            {
+                throw new ArgumentException("Values is null");
+            }
+        }
+
+        public ArrayList(T[] values)
+        {
+            if (values != null)
+            {
+                Length = values.Length;
+                _array = new T[(int)(values.Length * 2)];
+                for (int i = 0; i < Length; i++)
+                {
+                    _array[i] = values[i];
+                }
+            }
+            else
+            {
+                throw new ArgumentException("Array is null");
+            }
+        }
+
+        /// <summary>
+        /// Adds a value to the end of the list.
+        /// </summary>
+        /// <param value="value">Value to be added.</param>
+        public void Add(T value)
+        {
+            int lastIndex = Length;
+            AddByIndex(value, lastIndex);
+        }
+
+        /// <summary>
+        /// Adds a value to the beginning of the list.
+        /// </summary>
+        /// <param value="value">Value to be added.</param>
+        public void AddFirst(T value)
+        {
+            int index = 0;
+            AddByIndex(value, index);
+        }
+
+        /// <summary>
+        /// Adds a value at the specified index.
+        /// </summary>
+        /// <param value="value">Value to be added.</param>
+        /// <param index="index">Index by which the value is inserted.</param>
+        public void AddByIndex(T value, int index)
+        {
+            if (index >= 0 && index <= Length)
+            {
+                if (Length >= _array.Length)
+                {
+                    UpSize();
+                }
+
+                for (int i = Length; i > index; i--)
+                {
+                    _array[i] = _array[i - 1];
+                }
+
+                _array[index] = value;
+                ++Length;
+            }
+            else
+            {
+                throw new IndexOutOfRangeException();
+            }
+        }
+
+        /// <summary>
+        /// Remove the last item in the list.
+        /// </summary>
+        public void Remove()
+        {
+            int lastIndex = Length - 1;
+            RemoveByIndex(lastIndex);
```

Comment:
--Length

Reply:
then I will have an extra length reduction

Reply:
It was about replacement. Instead of using 143, 144 lines you can write RemoveByIndex(--Length);